In [1]:
# Original dataset: https://www.consumerfinance.gov/data-research/hmda/historic-data/?geo=ny&records=all-records&field_descriptions=labels

# 1) Iteration by chunks: Null values
    The dataset is massive so we will deal with it by chunks.
    We split the dataset in chunks of x size, then iterate by each chunk and finally we summarize the total column's null values.

In [2]:
import pandas as pd

In [3]:
# We force to show all the records
pd.set_option('display.max_rows', None) 

# We create a dictionary to keep all null values from the different chunks
total_null_values = None

# Iterate across the different chunks directly from the dataset
for chunk in pd.read_csv('C:/Users/nonox/hmda_2017_ny.csv', chunksize=100000):
    chunk_null_values = chunk.isnull().sum()
    
    if total_null_values is None:
        total_null_values = chunk_null_values
    else:
        total_null_values += chunk_null_values
    
    # Total null values per chunk
    print(f'Null values per column in this chunk:\n{chunk_null_values}\n')

# Total null values for the whole dataset
print(f'Total null values per column in the entire dataset:\n{total_null_values}')


C:\Users\nonox\AppData\Local\Temp\ipykernel_21348\1901377790.py:8: DtypeWarning: Columns (36,38,44,46) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv('C:/Users/nonox/hmda_2017_ny.csv', chunksize=100000):


Null values per column in this chunk:
as_of_year                             0
respondent_id                          0
agency_name                            0
agency_abbr                            0
agency_code                            0
loan_type_name                         0
loan_type                              0
property_type_name                     0
property_type                          0
loan_purpose_name                      0
loan_purpose                           0
owner_occupancy_name                   0
owner_occupancy                        0
loan_amount_000s                       0
preapproval_name                       0
preapproval                            0
action_taken_name                      0
action_taken                           0
msamd_name                          6174
msamd                               6167
state_name                             0
state_abbr                             0
state_code                             0
county_name        

C:\Users\nonox\AppData\Local\Temp\ipykernel_21348\1901377790.py:8: DtypeWarning: Columns (34,36,38,44,46,48) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv('C:/Users/nonox/hmda_2017_ny.csv', chunksize=100000):


Null values per column in this chunk:
as_of_year                             0
respondent_id                          0
agency_name                            0
agency_abbr                            0
agency_code                            0
loan_type_name                         0
loan_type                              0
property_type_name                     0
property_type                          0
loan_purpose_name                      0
loan_purpose                           0
owner_occupancy_name                   0
owner_occupancy                        0
loan_amount_000s                       0
preapproval_name                       0
preapproval                            0
action_taken_name                      0
action_taken                           0
msamd_name                          3421
msamd                               3418
state_name                             0
state_abbr                             0
state_code                             0
county_name        

C:\Users\nonox\AppData\Local\Temp\ipykernel_21348\1901377790.py:8: DtypeWarning: Columns (34,36,38,42,44,46,48) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv('C:/Users/nonox/hmda_2017_ny.csv', chunksize=100000):


Null values per column in this chunk:
as_of_year                             0
respondent_id                          0
agency_name                            0
agency_abbr                            0
agency_code                            0
loan_type_name                         0
loan_type                              0
property_type_name                     0
property_type                          0
loan_purpose_name                      0
loan_purpose                           0
owner_occupancy_name                   0
owner_occupancy                        0
loan_amount_000s                       0
preapproval_name                       0
preapproval                            0
action_taken_name                      0
action_taken                           0
msamd_name                         13801
msamd                              13791
state_name                             0
state_abbr                             0
state_code                             0
county_name        

C:\Users\nonox\AppData\Local\Temp\ipykernel_21348\1901377790.py:8: DtypeWarning: Columns (36,38,44,46,48) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv('C:/Users/nonox/hmda_2017_ny.csv', chunksize=100000):


Null values per column in this chunk:
as_of_year                             0
respondent_id                          0
agency_name                            0
agency_abbr                            0
agency_code                            0
loan_type_name                         0
loan_type                              0
property_type_name                     0
property_type                          0
loan_purpose_name                      0
loan_purpose                           0
owner_occupancy_name                   0
owner_occupancy                        0
loan_amount_000s                       0
preapproval_name                       0
preapproval                            0
action_taken_name                      0
action_taken                           0
msamd_name                          8870
msamd                               8866
state_name                             0
state_abbr                             0
state_code                             0
county_name        

C:\Users\nonox\AppData\Local\Temp\ipykernel_21348\1901377790.py:8: DtypeWarning: Columns (36,38,44,46) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv('C:/Users/nonox/hmda_2017_ny.csv', chunksize=100000):


# 2) Sampling for exploration:
For an easier exploration, sampled a tiny part of the whole.

In [4]:
df = pd.read_csv('C:/Users/nonox/hmda_2017_ny.csv')

# Sample
df_sample = df.sample(n=12000, random_state=42)  

# Save the dataset
df_sample.to_csv('C:/Users/nonox/hmda_2017_ny_sampled.csv', index=False)

print("Sampled dataset saved to 'C:/Users/nonox/hmda_2017_ny_sampled.csv'")


C:\Users\nonox\AppData\Local\Temp\ipykernel_21348\1043214581.py:1: DtypeWarning: Columns (34,36,38,44,46,48) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('C:/Users/nonox/hmda_2017_ny.csv')


Sampled dataset saved to 'C:/Users/nonox/hmda_2017_ny_sampled.csv'


In [5]:
pd.set_option('display.max_rows', None)
print(df_sample.describe())

       as_of_year   agency_code     loan_type  property_type  loan_purpose  \
count     12000.0  12000.000000  12000.000000   12000.000000  12000.000000   
mean       2017.0      6.879917      1.290000       1.034833      1.809417   
std           0.0      2.286540      0.581603       0.223512      0.927002   
min        2017.0      1.000000      1.000000       1.000000      1.000000   
25%        2017.0      7.000000      1.000000       1.000000      1.000000   
50%        2017.0      7.000000      1.000000       1.000000      1.000000   
75%        2017.0      9.000000      1.000000       1.000000      3.000000   
max        2017.0      9.000000      4.000000       3.000000      3.000000   

       owner_occupancy  loan_amount_000s   preapproval  action_taken  \
count     12000.000000      12000.000000  12000.000000  12000.000000   
mean          1.115833        346.848750      2.828250      2.518083   
std           0.354156       1396.116964      0.438484      1.854443   
min      

# 3) Iteration by chunks: "No co-applicants"
    a) First step in our data selection is to try to reduce the number of record as soon as possible.
    b) By focusing on "No co-aplicants" we significantly reduce the number of records.
    c) Our focus is on single individuals, we only take "No co-applicat" records.

In [6]:
# Initialize an empty DataFrame to store the filtered data
filtered_df = pd.DataFrame()

# Processing the CSV file in chunks
for chunk in pd.read_csv('C:/Users/nonox/hmda_2017_ny.csv', chunksize=100000):
    # Filter the chunk where "co_applicant_ethnicity_name" is "No co-applicant"
    filtered_chunk = chunk[chunk['co_applicant_ethnicity_name'] == 'No co-applicant']
    
    # Append the filtered chunk to the filtered_df DataFrame
    filtered_df = pd.concat([filtered_df, filtered_chunk])

# New CSV file
filtered_df.to_csv('C:/Users/nonox/hmda_2017_ny_no_co_applicant.csv', index=False)

print(f"Filtered dataset saved to '{'C:/Users/nonox/hmda_2017_ny_no_co_applicant.csv'}'")


#https://www.geeksforgeeks.org/how-to-create-an-empty-dataframe-and-append-rows-columns-to-it-in-pandas/

C:\Users\nonox\AppData\Local\Temp\ipykernel_21348\511449127.py:5: DtypeWarning: Columns (36,38,44,46) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv('C:/Users/nonox/hmda_2017_ny.csv', chunksize=100000):
C:\Users\nonox\AppData\Local\Temp\ipykernel_21348\511449127.py:5: DtypeWarning: Columns (34,36,38,44,46,48) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv('C:/Users/nonox/hmda_2017_ny.csv', chunksize=100000):
C:\Users\nonox\AppData\Local\Temp\ipykernel_21348\511449127.py:5: DtypeWarning: Columns (34,36,38,42,44,46,48) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv('C:/Users/nonox/hmda_2017_ny.csv', chunksize=100000):
C:\Users\nonox\AppData\Local\Temp\ipykernel_21348\511449127.py:5: DtypeWarning: Columns (36,38,44,46,48) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv('C:/User

Filtered dataset saved to 'C:/Users/nonox/hmda_2017_ny_no_co_applicant.csv'


In [7]:
# Checking new dimension:

pd.set_option('display.max_rows', None) 
df = pd.read_csv('C:/Users/nonox/hmda_2017_ny_no_co_applicant.csv')

# Shape
print(df.shape, '\n')

# Null values
print(df.isnull().sum())

C:\Users\nonox\AppData\Local\Temp\ipykernel_21348\2794338369.py:4: DtypeWarning: Columns (34,36,38) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('C:/Users/nonox/hmda_2017_ny_no_co_applicant.csv')


(242296, 78) 

as_of_year                             0
respondent_id                          0
agency_name                            0
agency_abbr                            0
agency_code                            0
loan_type_name                         0
loan_type                              0
property_type_name                     0
property_type                          0
loan_purpose_name                      0
loan_purpose                           0
owner_occupancy_name                   0
owner_occupancy                        0
loan_amount_000s                       0
preapproval_name                       0
preapproval                            0
action_taken_name                      0
action_taken                           0
msamd_name                         21138
msamd                              21120
state_name                             0
state_abbr                             0
state_code                             0
county_name                          455
c

# Random checking: 

 1) Statistics

In [8]:
pd.set_option('display.max_rows', None)
df = pd.read_csv('C:/Users/nonox/hmda_2017_ny_no_co_applicant.csv')
print(df.describe())

C:\Users\nonox\AppData\Local\Temp\ipykernel_21348\330022007.py:2: DtypeWarning: Columns (34,36,38) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('C:/Users/nonox/hmda_2017_ny_no_co_applicant.csv')


       as_of_year    agency_code      loan_type  property_type   loan_purpose  \
count    242296.0  242296.000000  242296.000000  242296.000000  242296.000000   
mean       2017.0       6.731539       1.255328       1.050591       1.852362   
std           0.0       2.329735       0.557420       0.277363       0.921090   
min        2017.0       1.000000       1.000000       1.000000       1.000000   
25%        2017.0       5.000000       1.000000       1.000000       1.000000   
50%        2017.0       7.000000       1.000000       1.000000       1.000000   
75%        2017.0       9.000000       1.000000       1.000000       3.000000   
max        2017.0       9.000000       4.000000       3.000000       3.000000   

       owner_occupancy  loan_amount_000s    preapproval   action_taken  \
count    242296.000000      2.422960e+05  242296.000000  242296.000000   
mean          1.143927      3.635772e+02       2.818255       2.261940   
std           0.394643      7.259435e+03       0

In [9]:

# 2. :

    # "property_type_name"

print(df['property_type_name'].value_counts())

property_type_name
One-to-four family dwelling (other than manufactured housing)    233539
Manufactured housing                                               5256
Multifamily dwelling                                               3501
Name: count, dtype: int64


# 5) Irrelevant columns:
    Apart from consider only records with "no_co-applicant"; We delete irrelevant columns


    'census_tract_number',                  
    'as_of_year',                          
    'respondent_id',
    'agency_name',                        
    'agency_abbr',                        
    'agency_code',                         
    'property_type',
    'owner_occupancy_name',
    'owner_occupancy',
    'preapproval_name',
    'preapproval',
    'state_name',
    'state_code',
    'state_abbr',
    'county_name',  
    'county_code',       
    'co_applicant_ethnicity_name',            
    'co_applicant_ethnicity',                                                         
    'applicant_race_2',
    'applicant_race_name_2',
    'applicant_race_name_3',             
    'applicant_race_3',             
    'applicant_race_name_4',
    'applicant_race_4',   
    'applicant_race_name_5',             
    'applicant_race_5',   
    'co_applicant_race_name_1',               
    'co_applicant_race_1',
    'co_applicant_race_name_2',
    'co_applicant_race_2', 
    'co_applicant_race_name_3', 
    'co_applicant_race_3',
    'co_applicant_race_name_4',
    'co_applicant_race_4',
    'co_applicant_race_4',               
    'co_applicant_race_name_5',          
    'co_applicant_race_5',               
    'co_applicant_sex_name',                  
    'co_applicant_sex',   
    'purchaser_type_name',
    'purchaser_type',                         
    'hoepa_status_name',                      
    'hoepa_status',
    'edit_status_name',                  
    'edit_status',
    'sequence_number',
    'population',
    'number_of_owner_occupied_units',       
    'number_of_1_to_4_family_units',        
    'application_date_indicator'


In [10]:
print(df['action_taken_name'].value_counts(), '\n')


print(df['denial_reason_name_1'].value_counts())

action_taken_name
Loan originated                                        132815
Application denied by financial institution             44972
Application withdrawn by applicant                      28299
Loan purchased by the institution                       14734
File closed for incompleteness                          11915
Application approved but not accepted                    9555
Preapproval request denied by financial institution         5
Preapproval request approved but not accepted               1
Name: count, dtype: int64 

denial_reason_name_1
Debt-to-income ratio                              10395
Credit history                                     8637
Collateral                                         5948
Credit application incomplete                      3859
Other                                              2813
Insufficient cash (downpayment, closing costs)      924
Unverifiable information                            905
Employment history                           

# 6) Filtering columns:

#### Looking at "action_taken" column:
###### Our target is approved/denied. Therefore we will keep only:

1) Loan originated                                        111541
2) Application denied by financial institution             32229
3) Application approved but not accepted                    9555

###### Hence, from "action_taken" column we will get rid off:
1) Application withdrawn by applicant                      28299
2) Loan purchased by the institution                       14734
3) File closed for incompleteness                          11915
4) Preapproval request denied by financial institution         5
5) Preapproval request approved but not accepted               1

#### From the column "denial_reason_name_1":
1) We will get rid of "Credit application incomplete"

In [11]:
df = df[(df['denial_reason_name_1'] != "Credit application incomplete") &

        (df['action_taken_name'] != "Application withdrawn by applicant") &
        (df['action_taken_name'] != "Loan purchased by the institution") & 
        (df['action_taken_name'] != "File closed for incompleteness") &
        (df['action_taken_name'] != "Preapproval request denied by financial institution") &
        (df['action_taken_name'] != 'Preapproval request approved but not accepted')]

# Random checking

In [12]:
print(df['action_taken_name'].value_counts())

action_taken_name
Loan originated                                132815
Application denied by financial institution     41113
Application approved but not accepted            9555
Name: count, dtype: int64


# 7) Approved vs Denied
    We rename and reorganize the different records we have on "action_taken_name"

In [13]:
df['action_taken'] = df['action_taken_name'].replace({
    'Loan originated': 'approved', 
    'Application approved but not accepted': 'approved',
    'Application denied by financial institution': 'denied'
})

In [14]:
print(df['action_taken'].value_counts())

action_taken
approved    142370
denied       41113
Name: count, dtype: int64


In [15]:
print("Column names in the dataset:")
print(df.columns)

Column names in the dataset:
Index(['as_of_year', 'respondent_id', 'agency_name', 'agency_abbr',
       'agency_code', 'loan_type_name', 'loan_type', 'property_type_name',
       'property_type', 'loan_purpose_name', 'loan_purpose',
       'owner_occupancy_name', 'owner_occupancy', 'loan_amount_000s',
       'preapproval_name', 'preapproval', 'action_taken_name', 'action_taken',
       'msamd_name', 'msamd', 'state_name', 'state_abbr', 'state_code',
       'county_name', 'county_code', 'census_tract_number',
       'applicant_ethnicity_name', 'applicant_ethnicity',
       'co_applicant_ethnicity_name', 'co_applicant_ethnicity',
       'applicant_race_name_1', 'applicant_race_1', 'applicant_race_name_2',
       'applicant_race_2', 'applicant_race_name_3', 'applicant_race_3',
       'applicant_race_name_4', 'applicant_race_4', 'applicant_race_name_5',
       'applicant_race_5', 'co_applicant_race_name_1', 'co_applicant_race_1',
       'co_applicant_race_name_2', 'co_applicant_race_2',
  

# 8) Feature Selection

In [16]:
df = df[

        ['loan_type_name',
        'property_type_name',
        'loan_purpose_name',
        'loan_amount_000s',
        'action_taken',
        'msamd_name',
        'applicant_ethnicity_name', 
        'applicant_race_name_1',
        'applicant_sex_name',
        'applicant_income_000s', 
        'denial_reason_name_1',
        'denial_reason_name_2',
        'denial_reason_name_3', 
        'rate_spread',
        'lien_status_name',
        'minority_population',
        'hud_median_family_income',
        'tract_to_msamd_income']

        ]

In [17]:
df.to_csv('C:/Users/nonox/hmdaNY27082024_2.csv', index=False)

print(f"Dataset saved to '{'C:/Users/nonox/hmdaNY27082024_2.csv'}'")

Dataset saved to 'C:/Users/nonox/hmdaNY27082024_2.csv'


    Original dataset: https://www.consumerfinance.gov/data-research/hmda/historic-data/?geo=ny&records=all-records&field_descriptions=labels
    
# So far from original dataset: 
1) Selected "no co-applicant" records from the column "co_applicant_ethnicity_name"
2) Filter columns:
    - action_taken_name
    - denial_reason_name_1
3) Created the column "action_taken":
    - df['action_taken'] = df['action_taken_name'].replace({
    'Loan originated': 'approved', 
    'Application approved but not accepted': 'approved',
    'Application denied by financial institution': 'denied'
})  
4) First feature selection:
    - df = df[

        ['loan_type_name',
        'property_type_name',
        'loan_purpose_name',
        'loan_amount_000s',
        'action_taken',
        'msamd_name',
        'applicant_ethnicity_name', 
        'applicant_race_name_1',
        'applicant_sex_name',
        'applicant_income_000s', 
        'denial_reason_name_1',
        'denial_reason_name_2',
        'denial_reason_name_3', 
        'rate_spread',
        'lien_status_name',
        'minority_population',
        'hud_median_family_income',
        'tract_to_msamd_income']

        ]
5) Excluded "Credit application incomplete" records from the column "denial_reason_name_1"
